src/workload.rs

```diff
@@ -144,32 +174,108 @@ impl Workload for DoubleInsertion {
+                Ok(())
+            }
+            .await;
+
+            if res.is_err() {
+                send!(conn, "ROLLBACK")?;
+            }
+            info!(log, "{:?}", res);
+
+            collect_result(res, results, &table, injection, pool.clone()).await;
+            disable_failpoint(
+                client,
+                "github.com/pingcap/tidb/table/tables/corruptMutations",
+            )
+            .await?;
+        }
+
+        Ok(())
+    }
+}
+
+// 2 txns. The first writes corrupted data and the second reads it. Check if Assertion can detect it.
+struct T2;
+#[async_trait]
+impl Workload for T2 {
+    async fn execute(
+        &self,
+        log: Logger,
+        config: &Config,
+        table: Table,
+        client: &reqwest::Client,
+        pool: Arc<Pool<MySql>>,
+        results: &mut HashMap<(Table, String, String), Effectiveness>,
+    ) -> Result<()> {
+        let mut conn = pool.acquire().await?;
+        enable_featuers(&mut conn, config).await?;
+        let drop_statement = table.drop_statement();
+        let create_statement = table.create_statement();
+        let start = Instant::now();
+        send!(conn, query(drop_statement.as_str())).expect("don't let drop statement fail");
+        send!(conn, query(create_statement.as_str())).expect("don't let create statement fail");
+        let duration = start.elapsed();
+        CREATE_TABLE_DURAION_MS.fetch_add(duration.as_millis() as u64, Ordering::SeqCst);
+        info!(log, "{}", create_statement);
+
+        for injection in AVAILABLE_INJECTIONS {
+            info!(log, "{} ready to go!", injection);
+
+            // NOTE: "1*" here, otherwise an index mutation is missing for each row insertion, thus cannot be detected.
+            enable_failpoint(
+                client,
+                "github.com/pingcap/tidb/table/tables/corruptMutations",
+                format!("1*return(\"{}\")", injection),
+            )
+            .await?;
+            send!(conn, "begin optimistic")?;
+            let row = table.new_row();
+            let insertion = format!("INSERT INTO {} VALUES ({})", table.name, row.to_string());
+            let deletion = format!(
+                "DELETE FROM {} WHERE {} = {}",
+                table.name,
+                table.cols[0].name,
+                row.cols[0].to_string()
+            );
+            let update = format!(
+                "UPDATE {} SET {} = {} WHERE {} = {}",
+                table.name,
+                table.cols[0].name,
+                row.cols[0].next().to_string(),
+                table.cols[1].name,
+                row.cols[1].to_string()
+            );
+
+            let res = async {
+                send!(conn, query(insertion.as_str()))?;
+                send!(conn, "commit")?;
+                send!(conn, "begin optimistic")?;
+                send!(conn, query(deletion.as_str()))?;
+                send!(conn, query(update.as_str()))?;
```

Comment:
Does this mean delete and then update? What's the purpose here?

Reply:
It's a mistake. Should be a delete after update.
I added the update stmt to see if it catches some cases of corruption that the deletion failed to detect.